File: cpp/leetcode/144.cpp
class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        stack<TreeNode*> stk;
        vector<int> res;
        if(!root) return res;
        stk.push(root);
        while(!stk.empty()) {
            TreeNode* node = stk.top();
            res.push_back(node->val);
            stk.pop();
            if(node->right) stk.push(node->right);
            if(node->left) stk.push(node->left);
        }
        return res;
    }
};


class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        if(!root) return {};
        vector<int> res;
        stack<TreeNode*> stk;
        while(root || !stk.empty()){
            if(root){
                res.push_back(root->val);
                stk.push(root);
                root = root->left;
            } else {
                TreeNode* node = stk.top(); stk.pop();
                root = node->right;
            }
        }
        return res;
    }
};

File: python/leetcode/144.py
class Solution:
    def help(self, root, res):
        if not root: return
        res.append(root.val)
        self.help(root.left, res)
        self.help(root.right, res)

    def preorderTraversal(self,root):
        res = []
        self.help(root, res)
        return res


class Solution(object):
    def preorderTraversal(self, root):
        res, stack = [], []
        while root or stack:
            if root:
                res.append(root.val)
                stack.append(root)
                root = root.left
            else:
                root = stack.pop()
                root = root.right
        return res
